src/libraries/System.Text.Json/src/System/Text/Json/Serialization/Converters/JsonDictionaryDefaultConverter.cs

```diff
@@ -0,0 +1,311 @@
+                    }
+
+                    // Determine the property.
+                    if (state.Current.PropertyState < StackFramePropertyState.Name)
+                    {
+                        if (reader.TokenType == JsonTokenType.EndObject)
+                        {
+                            break;
+                        }
+
+                        if (reader.TokenType != JsonTokenType.PropertyName)
+                        {
+                            ThrowHelper.ThrowJsonException_DeserializeUnableToConvertValue(TypeToConvert);
+                        }
+
+                        state.Current.PropertyState = StackFramePropertyState.Name;
+
+                        // Verify property doesn't contain metadata.
+                        if (shouldReadPreservedReferences)
+                        {
+                            ReadOnlySpan<byte> propertyName = JsonSerializer.GetSpan(ref reader);
+                            if (propertyName.Length > 0 && propertyName[0] == '$')
+                            {
+                                JsonSerializer.ThrowUnexpectedMetadataException(propertyName, ref reader, ref state);
+                            }
+                        }
+
+                        state.Current.JsonPropertyNameAsString = reader.GetString();
+                    }
+
+                    if (state.Current.PropertyState < StackFramePropertyState.ReadValue)
+                    {
+                        state.Current.PropertyState = StackFramePropertyState.ReadValue;
+
+                        if (!SingleValueReadWithReadAhead(elementConverter.ClassType, ref reader, ref state))
+                        {
+                            value = default!;
+                            return false;
+                        }
+                    }
+
+                    if (state.Current.PropertyState < StackFramePropertyState.TryRead)
+                    {
+                        // Read the value and add.
+                        bool success = elementConverter.TryRead(ref reader, typeof(TValue), options, ref state, out TValue element);
+                        if (!success)
+                        {
+                            value = default!;
+                            return false;
+                        }
+
+                        Add(element, options, ref state);
+                        state.Current.EndElement();
+                    }
+                }
+            }
+
+            ConvertCollection(ref state, options);
+            value = (TCollection)state.Current.ReturnValue!;
+            return true;
+        }
+
+        internal override sealed bool OnTryWrite(
+            Utf8JsonWriter writer,
+            TCollection dictionary,
+            JsonSerializerOptions options,
+            ref WriteStack state)
+        {
+            bool success;
+
+            if (dictionary == null)
+            {
+                writer.WriteNullValue();
+                success = true;
```

Comment:
nit: return true and remove the else so its easier to know that this finishes when the value is null.